`TestDatabaseRootLeafIdleTimeout/root_role_with_idle_timeout` flakiness

## Failure

#### Link(s) to logs

- https://console.cloud.google.com/cloud-build/builds/81ccd415-95ed-4224-ba5d-df9979643348?project=ci-account

#### Relevant snippet

```
    --- FAIL: TestDatabaseRootLeafIdleTimeout/root_role_with_idle_timeout (1.30s)
    helpers.go:260: event type "client.disconnect" not found after 1s
```

```
    helpers.go:281: event type "client.disconnect" not found after 1s
    --- FAIL: TestDatabaseRootLeafIdleTimeout/root_role_with_idle_timeout (1.51s)
```

Another instance: https://github.com/gravitational/teleport/actions/runs/5162756153/jobs/9300668031#step:6:1787

--- a/lib/srv/monitor.go
+++ b/lib/srv/monitor.go
@@ -353,13 +353,7 @@ func (w *Monitor) start(lockWatch types.Watcher) {
 						w.Entry.WithError(err).Warn("Failed to send idle timeout message.")
 					}
 				}
-				w.Entry.Debugf("Disconnecting client: %v", reason)
-				if err := w.Conn.Close(); err != nil {
-					w.Entry.WithError(err).Error("Failed to close connection.")
-				}
-				if err := w.emitDisconnectEvent(reason); err != nil {
-					w.Entry.WithError(err).Warn("Failed to emit audit event.")
-				}
+				w.disconnectClient(reason)
 				return
 			}
 			next := w.ClientIdleTimeout - since
@@ -408,14 +402,18 @@ func (w *Monitor) start(lockWatch types.Watcher) {
 
 func (w *Monitor) disconnectClientOnExpiredCert() {
 	reason := fmt.Sprintf("client certificate expired at %v", w.Clock.Now().UTC())
+	w.disconnectClient(reason)
+}
 
+func (w *Monitor) disconnectClient(reason string) {
 	w.Entry.Debugf("Disconnecting client: %v", reason)
-	if err := w.Conn.Close(); err != nil {
-		w.Entry.WithError(err).Error("Failed to close connection.")
-	}
-
+	// Emit Audit event first to make sure that that underlying context will not be canceled during
+	// emitting audit event.
 	if err := w.emitDisconnectEvent(reason); err != nil {
 		w.Entry.WithError(err).Warn("Failed to emit audit event.")
+	}
+	if err := w.Conn.Close(); err != nil {
+		w.Entry.WithError(err).Error("Failed to close connection.")
 	}
 }
 
@@ -448,13 +446,7 @@ func (w *Monitor) handleLockInForce(lockErr error) {
 			w.Entry.WithError(err).Warn("Failed to send lock-in-force message.")
 		}
 	}
-	w.Entry.Debugf("Disconnecting client: %v.", reason)
-	if err := w.Conn.Close(); err != nil {
-		w.Entry.WithError(err).Error("Failed to close connection.")
-	}
-	if err := w.emitDisconnectEvent(reason); err != nil {
-		w.Entry.WithError(err).Warn("Failed to emit audit event.")
-	}
+	w.disconnectClient(reason)
 }
 
 type trackingChannel struct {

